api-extractor/src/definitions/ApiItem.ts

```diff
@@ -399,6 +417,139 @@ abstract class ApiItem {
     }
     return false;
   }
+
+  /**
+   * This is called by ApiItems to visit the types that appear in an expression.  For example,
+   * if a Public API function returns a class that is defined in this package, but not exported,
+   * this is a problem. visitTypeReferencesForNode() finds all TypeReference child nodes under the
+   * specified node and analyzes each one.
+   */
+  protected visitTypeReferencesForNode(node: ts.Node): void {
+    if (node.kind === ts.SyntaxKind.Block ||
+      (node.kind >= ts.SyntaxKind.JSDocTypeExpression && node.kind <= ts.SyntaxKind.JSDocNeverKeyword)) {
+      // Don't traverse into code blocks or JSDoc items; we only care about the function signature
+      return;
+    }
+
+    if (node.kind === ts.SyntaxKind.TypeReference) {
+      const typeReference: ts.TypeReferenceNode = node as ts.TypeReferenceNode;
+      this._analyzeTypeReference(typeReference);
+    }
+
+    // Recurse the tree
+    for (const childNode of node.getChildren()) {
+      this.visitTypeReferencesForNode(childNode);
+    }
+  }
+
+  /**
+   * This is a helper for visitTypeReferencesForNode().  It analyzes a single TypeReferenceNode.
+   */
+  private _analyzeTypeReference(typeReferenceNode: ts.TypeReferenceNode): void {
+    const symbol: ts.Symbol = this.extractor.typeChecker.getSymbolAtLocation(typeReferenceNode.typeName);
+    if (!symbol) {
+      // Is this bad?
+      return;
+    }
+
+    if (symbol.flags & ts.SymbolFlags.TypeParameter) {
+      // Don't analyze e.g. "T" in "Set<T>"
+      return;
+    }
+
+    // Follow the aliases all the way to the ending SourceFile
+    const currentSymbol: ts.Symbol = this.followAliases(symbol);
+
+    if (!currentSymbol.declarations || !currentSymbol.declarations.length) {
+      // This is a degenerate case that happens sometimes
+      return;
+    }
+    const sourceFile: ts.SourceFile = currentSymbol.declarations[0].getSourceFile();
+
+    // Walk upwards from that directory until you find a directory containing package.json,
+    // this is where the referenced type is located.
+    // Example: "c:\users\<username>\sp-client\spfx-core\sp-core-library"
+    const typeReferencePackagePath: string = PackageJsonHelpers.tryFindPackagePathUpwards(sourceFile.path);
+    // Example: "@microsoft/sp-core-library"
+    let typeReferencePackageName: string = '';
+
+    // If we can not find a package path, we consider the type to be part of the current project's package. 
+    // One case where this happens is when looking for a type that is a symlink
+    if (!typeReferencePackagePath) {
+      typeReferencePackageName = this.extractor.package.name;
+    } else {
+      typeReferencePackageName = PackageJsonHelpers.readPackageName(typeReferencePackagePath);
```

Comment:
This is the place where we would want to apply a mapping from e.g. "@types/react" --> "react", to ensure that the DocItemLoader doesn't get confused when it tries to load e.g. "react.api.json".

I'd suggest a simple hard-coded constant for now, something like this:

```javascript
// Names of NPM scopes that contain packages that provide typings for the real package.
// The TypeScript compiler's typings design doesn't seem to handle scoped NPM packages,
// so the transformation will always be simple, like this:
// "@types/example" --> "example"
// NOT like this:
// "@types/@contoso/example" --> "@contoso/example"
// "@contosotypes/example" --> "@contoso/example"
// Eventually this constant should be provided by the gulp task that invokes the compiler.
const typingsScopeNames: string[] = [ '@types' ]
```